filters/openpolicyagent/openpolicyagent.go

```diff
@@ -516,6 +596,102 @@ func (opa *OpenPolicyAgentInstance) Start(ctx context.Context, timeout time.Dura
 	return nil
 }
 
+// Start starts the policy engine's plugin manager and periodically trigger the plugins to download policies etc.
+func (opa *OpenPolicyAgentInstance) StartAndTriggerPlugins(ctx context.Context) error {
+	err := opa.manager.Start(ctx)
+
+	if err == nil {
+		ctxWithStartupTimeout, cancel := context.WithTimeout(ctx, opa.instanceConfig.startupTimeout)
+		defer cancel()
+		err = opa.triggerPluginsWithRetry(ctxWithStartupTimeout)
+	}
+	if err == nil {
+		err = opa.verifyAllPluginsStarted()
+	}
+
+	if err != nil {
+		opa.Close(ctx)
+		return err
+	}
+	return nil
+}
+
+func (opa *OpenPolicyAgentInstance) triggerPluginsWithRetry(ctx context.Context) error {
+	var httpError download.HTTPError
+	var err error
+	backoff := 100 * time.Millisecond
+	retry := 0
+
+	for {
+		select {
+		case <-ctx.Done():
+			return fmt.Errorf("context cancelled while triggering plugins: %w, last retry returned: %w", ctx.Err(), err)
+		default:
+			err = opa.triggerPlugins(ctx)
+
+			if ok := asRetryableHttpError(err, &httpError); !ok {
+				return err
+			}
+
+			opa.Logger().WithFields(map[string]interface{}{
+				"retry":  retry,
+				"status": httpError.StatusCode,
+			}).Warn("Triggering bundles failed. Retrying %d", retry)
+
+			time.Sleep(backoff)
+			backoff *= 2
+			retry++
+		}
+	}
+}
+
+func asRetryableHttpError(err error, httpError *download.HTTPError) bool {
+	isHttpError := errors.As(err, httpError)
+	if isHttpError && (httpError.StatusCode == 429 || httpError.StatusCode >= 500) {
+		return true
+	}
+	return false
+}
+
+func (opa *OpenPolicyAgentInstance) verifyAllPluginsStarted() error {
+	allPluginsReady := true
+	for pluginName, status := range opa.manager.PluginStatus() {
```

Comment:
Maybe we make it explicit that we only wait for the bundle and discovery plugin?

Reply:
it is checking the status of all plugins to align with the normal startup without the control loop